### PR TITLE
Reuters graph floor

### PR DESCRIPTION
**Pull request by fvlieshout**
I added the graph code to the train.py file that was created by @RedHilarious.
The original file (graph_train.py) is still in there just to be sure. I think we should let that one in there until we are very sure we don't need code from there anymore.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (Matyasch) [inline .vscode/settings.json]: This file shouldn't be in the repo

File: .gitignore [hunk 2]
```diff
@@ -1,7 +1,19 @@
+# Own files
+output/
```
Comment (Matyasch) [inline .gitignore:2]: This might not be a good idea if we want to save logs but ok

File: data_prep/graph_utils.py [hunk 3]
```diff
@@ -71,7 +71,7 @@ def word_pair_frequency(windows):
     return word_pair_freq
 
 
-def pmi(corpus, window_size=15):
+def PMI(corpus, window_size=15):
```
Comment (Matyasch) [inline data_prep/graph_utils.py]: rip https://www.python.org/dev/peps/pep-0008/#function-and-variable-names :(

File: datasets/reuters_graph_datasets.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
+from collections import defaultdict
```
Comment (Matyasch) [inline datasets/reuters_graph_datasets.py]: This file shouldn't exist

File: env.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (Matyasch) [inline env.yml:1]: This file shouldn't exist

File: models/model.py [hunk 6]
```diff
@@ -118,3 +121,66 @@ def forward(self, inputs, attention_mask=None):
         cls_token_state = hidden_states[1]
 
         return cls_token_state
+
+class Net(torch.nn.Module):
```
Comment (Matyasch) [inline models/model.py]: This name is too generic if this file is not purely for GNNs

File: reuters_graph.ipynb [hunk 7]
```diff
@@ -16,22 +15,223 @@
   {
    "cell_type": "code",
    "execution_count": 1,
-   "id": "5373b31a",
    "metadata": {},
+<<<<<<< HEAD
```
Comment (Matyasch) [inline reuters_graph.ipynb:19]: rip this notebook :(

File: models/model.py [hunk 8]
```diff
@@ -118,3 +121,66 @@ def forward(self, inputs, attention_mask=None):
...
+
+    def forward(self, data):
+        x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr
+        x = self.conv1(x, edge_index, edge_weight)
+        x = F.relu(x)
+        x = F.dropout(x, training=self.training)
+        x = self.conv2(x, edge_index, edge_weight)
+        return x
+
+class Graph_model(pl.LightningModule):
```
Comment (Matyasch) [inline models/model.py]: This should be "GraphModel"

File: data_prep/reuters_graph_datasets.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    
+    def len(self):
+        length = len(self.data)
+        return 1
+    
+    def get(self, idx):
+        return self.data
+
+
+class R52(Reuters):
```
Comment (Matyasch) [inline data_prep/reuters_graph_datasets.py]: This will have to be changed to R52Graph

File: data_prep/reuters_graph_datasets.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    """
+    Wrapper for the R52 dataset.
+    """
+    def __init__(self, device, val_size=0.1):
+        super().__init__(r8=False, device=device, val_size=val_size)
+
+
+class R8(Reuters):
```
Comment (Matyasch) [inline data_prep/reuters_graph_datasets.py]: This will have to be changed to R8Graph

File: data_prep/reuters_graph_datasets.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
+from collections import defaultdict
```
Comment (Matyasch) [inline data_prep/reuters_graph_datasets.py:1]: Is this a new file for safety reasons?

File: graph_train.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    def test_step(self, batch, batch_idx):
+        loss, acc = self.forward(batch, mode='test')
+        self.log("test_loss", loss)
+        self.log("test_acc", acc)
+
+class GenerateCallback(pl.Callback):
+    def __init__(self):
+        """
+        Args:
+            gamma (float, optional): [description]. Defaults to 0.2.
```
Comment (Matyasch) [inline graph_train.py:87]: gamma?

File: graph_train.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        """
+        super().__init__()
+
+    def on_epoch_end(self, trainer, pl_module):
+        """
+        This function is called after every epoch.
+        In this function, the dev accuracy is checked and the learning rate is adjusted if dev accuracy has decreased
```
Comment (Matyasch) [inline graph_train.py:94]: Ah this was copied from deep learning I guess. It would be better if it had a descriptive name and real (or no) docstring

File: graph_train.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+    def forward(self, data):
+        x, edge_index, edge_weight = data.x, data.edge_index, data.edge_attr
+        x = self.conv1(x, edge_index, edge_weight)
+        x = F.relu(x)
+        x = F.dropout(x, training=self.training)
+        x = self.conv2(x, edge_index, edge_weight)
+        return x
+
+class Graph_model(pl.LightningModule):
```
Comment (Matyasch) [inline graph_train.py:39]: Isn't this also in model.py?

File: train.py [hunk 15]
```diff
@@ -38,6 +40,7 @@ def train(model_name, seed, epochs, patience, b_size, l_rate, l_decay, minimum_l
     # the data preprocessing per model
 
     dataset = get_dataset(dataset_name)
+    optimizer_hparams = {"lr": l_rate, "weight_decay": l_decay}
```
Comment (Matyasch) [inline train.py]: Either this should say "learning_rate_decay" or the parameter should be called "w_decay"

File: train.py [hunk 16]
```diff
@@ -168,18 +182,22 @@ def on_validation_end(self, trainer, pl_module):
...
 
-    parser.add_argument('--epochs', dest='epochs', type=int, default=40)
+    parser.add_argument('--epochs', dest='epochs', type=int, default=20)
     parser.add_argument('--patience', dest='patience', type=int, default=10)
-    parser.add_argument('--batch-size', dest='batch_size', type=int, default=64)
-    parser.add_argument('--lr', dest='l_rate', type=float, default=1e-4)
-    parser.add_argument("--min-lr", dest='minimum_lr', type=float, default=1e-5, help="Minimum Learning Rate")
-    parser.add_argument("--lr-decay", dest='lr_decay', type=float, default=1e-3, help="Learning rate (weight) decay")
+    #batch_size 1 for graph networks
+    # parser.add_argument('--batch-size', dest='batch_size', type=int, default=64)
```
Comment (Matyasch) [inline train.py]: This shouldn't be commented out, we can just set it to 1 if we want to work with GNNs

File: train.py [hunk 17]
```diff
@@ -168,18 +182,22 @@ def on_validation_end(self, trainer, pl_module):
...
+    parser.add_argument('--batch-size', dest='batch_size', type=int, default=1)
+    #higher learning rate for graph models
+    # parser.add_argument('--lr', dest='l_rate', type=float, default=1e-4)
```
Comment (Matyasch) [inline train.py]: Why comment this out and have almost the same thing under it

File: train.py [hunk 18]
```diff
@@ -168,18 +182,22 @@ def on_validation_end(self, trainer, pl_module):
...
+    parser.add_argument('--lr', dest='l_rate', type=float, default=0.01)
+    parser.add_argument("--min-lr", dest='minimum_lr', type=float, default=0.001, help="Minimum Learning Rate")
+    parser.add_argument("--lr-decay", dest='lr_decay', type=float, default=0.01, help="Learning rate (weight) decay")
 
     # CONFIGURATION
 
-    parser.add_argument('--dataset', dest='dataset', default='R8Text', choices=SUPPORTED_DATASETS,
+    parser.add_argument('--dataset', dest='dataset', default='R8Graph', choices=SUPPORTED_DATASETS,
```
Comment (Matyasch) [inline train.py:194]: R8Graph is not a supported dataset in SUPPORTED_DATASETS

File: train.py [hunk 19]
```diff
@@ -168,18 +182,22 @@ def on_validation_end(self, trainer, pl_module):
...
-    parser.add_argument("--min-lr", dest='minimum_lr', type=float, default=1e-5, help="Minimum Learning Rate")
-    parser.add_argument("--lr-decay", dest='lr_decay', type=float, default=1e-3, help="Learning rate (weight) decay")
+    #batch_size 1 for graph networks
+    # parser.add_argument('--batch-size', dest='batch_size', type=int, default=64)
+    parser.add_argument('--batch-size', dest='batch_size', type=int, default=1)
+    #higher learning rate for graph models
+    # parser.add_argument('--lr', dest='l_rate', type=float, default=1e-4)
+    parser.add_argument('--lr', dest='l_rate', type=float, default=0.01)
+    parser.add_argument("--min-lr", dest='minimum_lr', type=float, default=0.001, help="Minimum Learning Rate")
+    parser.add_argument("--lr-decay", dest='lr_decay', type=float, default=0.01, help="Learning rate (weight) decay")
```
Comment (Matyasch) [inline train.py]: Not sure if overwriting default values (instead of just using the arguments when the script is called) is the cleanest thing to do

File: train.py [hunk 20]
```diff
@@ -168,18 +182,22 @@ def on_validation_end(self, trainer, pl_module):
...
 
     # CONFIGURATION
 
-    parser.add_argument('--dataset', dest='dataset', default='R8Text', choices=SUPPORTED_DATASETS,
+    parser.add_argument('--dataset', dest='dataset', default='R8Graph', choices=SUPPORTED_DATASETS,
                         help='Select the dataset you want to use.')
-    parser.add_argument('--model', dest='model', default='roberta', choices=SUPPORTED_MODELS,
+    parser.add_argument('--model', dest='model', default='pure_gnn', choices=SUPPORTED_MODELS,
```
Comment (Matyasch) [inline train.py:196]: pure_gnn is not a supported model in SUPPORTED_MODELS

File: train_GNN.job [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+module purge
+module load 2019
+module load Python/3.7.5-foss-2019b
+module load CUDA/10.1.243
+module load cuDNN/7.6.5.32-CUDA-10.1.243
+module load NCCL/2.5.6-CUDA-10.1.243
+module load Anaconda3/2018.12
+
+# Your job starts in the directory where you call sbatch
+cd /home/lcur0321/ATCS_group1
```
Comment (Matyasch) [inline train_GNN.job]: This should be $HOME or it breaks on any other account

**Review by Matyasch — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
Comment (fvlieshout) [reply .gitignore:2]: I only save the 'prints' from Lisa in this folder. Usually I have a lot of shitty/not working output files in here which don't make sense to put on github

Comment (Matyasch) [reply .gitignore:2]: Sure but the .gitignore file also applies to us. You should use https://stackoverflow.com/a/5724499 instead

File: data_prep/graph_utils.py (hunk 3 above)
Comment (fvlieshout) [reply data_prep/graph_utils.py]: it didn't match with the other files anymore...

File: data_prep/reuters_graph_datasets.py (hunk 11 above)
Comment (fvlieshout) [reply data_prep/reuters_graph_datasets.py:1]: yes indeed. In the end, we can delete the other file

File: data_prep/reuters_graph_datasets.py (hunk 9 above)
Comment (fvlieshout) [reply data_prep/reuters_graph_datasets.py:195]: Okay, I copied this from the other Reuters_Graph python file. Did you make any adaptions to R52 there already that are now missing here? In that case, we can just add them here as well

File: data_prep/reuters_graph_datasets.py (hunk 10 above)
Comment (fvlieshout) [reply data_prep/reuters_graph_datasets.py]: Done

File: data_prep/reuters_graph_datasets.py (hunk 9 above)
Comment (fvlieshout) [reply data_prep/reuters_graph_datasets.py]: Done

File: env.yml (hunk 5 above)
Comment (fvlieshout) [reply env.yml:1]: Why not?

File: graph_train.py (hunk 14 above)
Comment (fvlieshout) [reply graph_train.py:39]: Yes this file should also be deleted in the end. Maybe for this file, we can already do it right now as we can always track it back if we need to. do you agree on deleting?

File: graph_train.py (hunk 12 above)
Comment (fvlieshout) [reply graph_train.py:87]: This was copied from my first assignment. File should be deleted anyways

File: graph_train.py (hunk 13 above)
Comment (fvlieshout) [reply graph_train.py:94]: Same as above. Copied and should be deleted

File: models/model.py (hunk 6 above)
Comment (fvlieshout) [reply models/model.py]: Changed it

File: train.py (hunk 15 above)
Comment (fvlieshout) [reply train.py]: This was copied from Rahel's code but I indeed changed the parameter to w_decay

File: train.py (hunk 16 above)
Comment (fvlieshout) [reply train.py]: Changed it

File: train.py (hunk 17 above)
Comment (fvlieshout) [reply train.py]: Changed it

File: train.py (hunk 19 above)
Comment (fvlieshout) [reply train.py]: Where are we overwriting the default values?

File: train.py (hunk 18 above)
Comment (fvlieshout) [reply train.py:194]: I added it (and also R52)

File: train.py (hunk 20 above)
Comment (fvlieshout) [reply train.py:196]: I added it

File: train_GNN.job (hunk 21 above)
Comment (fvlieshout) [reply train_GNN.job]: how exactly?

File: env.yml (hunk 5 above)
Comment (Matyasch) [reply env.yml:1]: Because the datasets directory does not exist anymore

File: data_prep/graph_utils.py (hunk 3 above)
Comment (Matyasch) [reply data_prep/graph_utils.py]: Hm that's a problem, but then the other files should be fixed instead

File: graph_train.py (hunk 14 above)
Comment (Matyasch) [reply graph_train.py:39]: Hm it's fine for now I guess. Maybe next time when we do this we could add a comment on top of the file like "THIS FILE IS A DUPLICATE AND WILL BE DELETED" or something. Also to not make our future code depend on it if someone does not know our plans with the file

File: train.py (hunk 19 above)
Comment (Matyasch) [reply train.py]: You changed the default value of:

- epochs from 40 to 20
- batch size from 64 to 1
- learning rate from 1e-4 to 0.01
- min learning rate from 1e-5 to 0.001
- learning rate decay from 1e-3 to 0.01
I assume these values might have been important for Roberta

File: train_GNN.job (hunk 21 above)
Comment (Matyasch) [reply train_GNN.job]: I cannot access `/home/lcur0321/ATCS_group1` from my account `lcur0330`.
```bash
lcur0330@login3:$ cd /home/lcur0321/ATCS_group1
-bash: cd: /home/lcur0321/ATCS_group1: Permission denied
```

So you should use `cd $HOME/ATCS_group1` instead because `$HOME` always maps to the current user's home directory
```bash
lcur0330@login3:~/matyas/ATCS_group1$ $HOME
-bash: /home/lcur0330: Is a directory
```

File: data_prep/graph_utils.py (hunk 3 above)
Comment (fvlieshout) [reply data_prep/graph_utils.py]: Fixed it

File: graph_train.py (hunk 14 above)
Comment (fvlieshout) [reply graph_train.py:39]: haha good idea. I added it

File: train.py (hunk 19 above)
Comment (fvlieshout) [reply train.py]: yuo're right. I changed this to run it on my own machine for debugging but I now changed them back

File: train_GNN.job (hunk 21 above)
Comment (fvlieshout) [reply train_GNN.job]: Perfect, thanks. Changed it

**Review by Matyasch — APPROVED** [review]
